### ISSUE TITLE
Memory usage - opening/saving files

### ISSUE THREAD

**Issue by clumba** [ClosedXML/ClosedXML#1280]
## Read and complete the full issue template

**Do you want to request a *feature* or report a *bug*?**
- [x] Bug
- [ ] Feature

**Version of ClosedXML**

0.94.2

**What is the current behavior?**

During opening and especially saving files memory usage spikes high even if nothing was edited. Is this normal? My work files is somewhat big - 20k rows, but not that big for a 200/300+ MB RAM for saving process.

[upd] I don't know why, but I can't upload my file, so I used random big xlsx from data.gov: https://data.gov.au/data/dataset/historical-australian-government-contract-data/resource/c7983dca-5286-4e88-8de4-bc421dd156fe

It bigger than my example, but there is same picture.

Also as you can see .Dispose() not freeing memory like wb = null. Seems like a bug.

**What is the expected behavior or new feature?**

Can it be lower? 

**Did this work in previous versions of our tool?  Which versions?**

Yes

## Reproducibility

**Code to reproduce problem:**
```c#
using ClosedXML.Excel;
using System;
using System.IO;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace closedxml_memory_usage
{
    class Program
    {
        static void Main(string[] args)
        {
            Process proc = Process.GetCurrentProcess();
            Console.WriteLine("Program started.");
            Console.WriteLine("Memory usage, MB: " + proc.PrivateMemorySize64 / (1024*1024));
            Console.WriteLine();

            var wb = new XLWorkbook(@"d:\20142015fy (1).xlsx");
            proc.Refresh();
            Console.WriteLine("XLWorkbook opened file 12.7MB, 69k rows");
            Console.WriteLine("Memory usage, MB: " + proc.PrivateMemorySize64 / (1024 * 1024));
            Console.WriteLine();


            wb.SaveAs(@"d:\file1.xlsx");
            Console.WriteLine("XLWorkbook file 12.7MB, 69k rows");
            proc.Refresh();
            Console.WriteLine("Memory usage, MB: " + proc.PrivateMemorySize64 / (1024 * 1024));
            Console.WriteLine();


            GC.Collect();
            GC.WaitForPendingFinalizers();
            Console.WriteLine("Running GC");
            proc.Refresh();
            Console.WriteLine("Memory usage, MB: " + proc.PrivateMemorySize64 / (1024 * 1024));
            Console.WriteLine();


            wb.Dispose();
            GC.Collect();
            GC.WaitForPendingFinalizers();
            Console.WriteLine("wb.dispose() + Running GC");
            proc.Refresh();
            Console.WriteLine("Memory usage, MB: " + proc.PrivateMemorySize64 / (1024 * 1024));
            Console.WriteLine();

            wb = null;
            GC.Collect();
            GC.WaitForPendingFinalizers();
            Console.WriteLine("wb=null + Running GC");
            proc.Refresh();
            Console.WriteLine("Memory usage, MB: " + proc.PrivateMemorySize64 / (1024 * 1024));
            Console.ReadLine();
        }
    }
}


```
- [ ] I attached a sample spreadsheet.  (You can drag files on to this issue)
Used file:

https://data.gov.au/data/dataset/historical-australian-government-contract-data/resource/c7983dca-5286-4e88-8de4-bc421dd156fe

My results:

```
Program started.
Memory usage, MB: 10

XLWorkbook opened file 12.7MB, 69k rows
Memory usage, MB: 308

XLWorkbook saved file 12.7MB, 69k rows
Memory usage, MB: 1098

Running GC
Memory usage, MB: 299

wb.dispose() + Running GC
Memory usage, MB: 299

wb=null + Running GC
Memory usage, MB: 35

```

My file:

```
Program started.
Memory usage, MB: 10

XLWorkbook opened file 1.69MB, 20k rows
Memory usage, MB: 75

XLWorkbook saved file 1.69MB.
Memory usage, MB: 206

Running GC
Memory usage, MB: 79

wb.dispose() + Running GC
Memory usage, MB: 80

wb=null + Running GC
Memory usage, MB: 29
```

**Comment by Pankraty** [comment]
Check !1276. It looks very similar

**Comment by clumba** [comment]
> Check #1276. It looks very similar

```
Program started.
Memory usage, MB: 10

XLWorkbook opened file 12.7MB, 69k rows
Memory usage, MB: 308

XLWorkbook saved 12.7MB
Memory usage, MB: 1099

Running GC
Memory usage, MB: 300

wb.dispose() + Running GC
Memory usage, MB: 46

wb=null + Running GC
Memory usage, MB: 36
```


Well, wb.dispose() seems working now. So, this memory usage is ok?

**Comment by Pankraty** [comment]
Let's check.
The file you provided has 69 237 and 34 columns which give us 2 354 058 cells. `XLCell` instances usually consume most of the memory so we can omit the rest for simplicity. Memory consumption is 308 MB, or approximately 322 961 408 bytes. By dividing one to another we'll get 137 bytes per cell at average. Each cell stores its value (a unicode string), data type, a reference to a style, address, "last modified at" and "cached at" counters and many other properties so 137 bytes do not seem too unreasonable. I wish it would be lower but there is not much we can do with the data itself anyway.
The main question is, whether we can reduce memory usage during saving. This will require investigation and if you could do some profiling to determine what causes it that would be very helpful.

**Comment by clumba** [comment]
> The main question is, whether we can reduce memory usage during saving. This will require investigation and if you could do some profiling to determine what causes it that would be very helpful.

Yes, this is my initial concern. What can I do to help you investigate this?

**Comment by Pankraty** [comment]
To start from, ClosedXML internally uses OpenXML so it would be nice to know how much of that memory consumption is caused by the OpenXML library. This is a background level which we hardly will be able to achieve. I don't expect high values there but still, since it processes XML a lot, there might be plenty of string objects.

Then, you can profile the XLWorkbook.Save method, setting breakpoints here and there and getting a memory dump on each of them. This will help to narrow down the search and will point at the steps causing most intensive allocations. After that, we can try to figure out how to reduce or avoid them.

I do not mind doing it myself, I am just a bit busy lately.

**Comment by clumba** [comment]
I'm afraid it's beyond my c# programming skills :(

**Comment by Pankraty** [comment]
What a surprise. A single line bloats memory usage from 556 MB to 1292 MB! And the line is
```
if (worksheetPart.Worksheet == null)
```
o_O

Writing sheet data adds 272 MB and generating calculation chains - 147 MB. The later looks weird as the workbook does not have a sinlge formula, and probably this part may be improved. As for the others - I don't think it will be easy to change.

**Comment by igitur** [comment]
Fixed by #1284
